### ISSUE TITLE
How to use with dynamic content (e.g. with Gatsby/React)?

### ISSUE THREAD

**Issue by ooloth** [Munter/subfont#29]
Thanks for creating this!

I'm encountering an issue when using subfont with React (rendered statically via GatsbyJS). I'm running subfont on the static HTML generated by Gatsby by running `gatsby build && subfont public -i`.

Subfont works perfectly for any content included in the initial HTML. However, for content rendered later based on user interaction (e.g. showing a menu, viewing more list items, etc.), any new letters are missing from the font and do not render properly.

As a temporary fix, I've added a hidden `div` at the end of the built HTML that includes the missing letters, but this is obviously hacky and not a workable solution in some cases (e.g. if users were to add new dynamic content via a CMS).

Is there an intended way to use subfont with dynamic content (i.e. content that is not included in the initial HTML build)? There's excitement in the Gatsby community about using subfont (e.g. [this tweet](https://twitter.com/kylemathews/status/998543855397015552) from @kylemathews), so I'd love to understand how to get around this issue.

**Comment by papandreou** [comment]
You're welcome, and thanks for taking the time to write.

At present, subfont does not execute any JavaScript and thus makes no attempt to trace dynamically injected stylesheets or HTML. As you point out, only the "statically detectable" content is taken into consideration. We should probably put that in capital letters in the readme.

However, great care is taken to make sure that the browser will load and render the original font if the subset does not contain all the required characters. If you see letters missing and text rendering incorrectly, that's certainly a bug. Are you able to share the output of gatsby that you're running `subfont public -i` on, then we'll look into that?

I'm happy that you still gave the tool a chance and were able to work around the missing capability by injecting that hidden `div`. We have been discussing the possibility of expanding the scope of the project so that we also run JavaScript. Here's a quick brain dump:

* Since assetgraph already uses jsdom internally, we could try to turn on the mode where it fetches and executes JavaScript.
* Alternatively, we could port the tracing part of the code to work in a browser, and then inject it into eg. Headless Chrome, pulling out the required information.
* ... Or we could just pull out the serialized document and all stylesheets, then do the tracing outside the headless browser
* ... Or we could cut a corner and just use `getComputedStyle` in the headless browser (like [glyphhanger](https://github.com/filamentgroup/glyphhanger))
* In all of the above scenarios, it probably won't be sufficient to only trace the final "on load" state of the DOM, as there might be intermediate states, "loading" screens etc. where text is rendered -- which should be included in the computed subsets. We could use something like `MutationObserver` for detecting when to run additional traces.

I have a preference for solutions that don't compromise too much wrt. the current feature set, so that we're also able to account for both possibilities when font-related properties are being changed in media queries, `<noscript>` blocks, multi-valued custom properties, conditional comments, etc. That makes `getComputedStyle` less appealing.

We should also entertain the idea of making the components of subfont more loosely coupled so that users could maybe gather text/font prop information from multiple sources via separate tools/plugins, then ask subfont to create subsets based on a synthesis of all those traces.

All of the above would take a bunch of time and effort to get done. Before we get too carried away, I think we should think about whether there's something simpler we can do for your particular use case in the mean time. Can you think of any cli switches or implicit configuration that would be convenient to work with for you, as an alternative to that hidden `div`?

**Comment by Munter** [comment]
One simple solution could be that I inject a `<link rel="prefetch">` for the original font, so it will be loaded as soon as the browser finds it appropriate during/after page load. That would very likely make the original font available already at the time when a user interacts with the page and causes these dynamic changes to happen.

Gatsby might actually be a special case where it could potentially be possible to extract text nodes during a build at he time where the vdom is available. That information could then be passed on to subfont with an instruction that these characters are known. This feature doesn't exist yet, but it would be a fairly easy addition to add an input of characters that must exist in all fonts. It becomes more difficult when it's about targeting specific characters for specific fonts, but that's also where the performance gains are to be found

**Comment by ooloth** [comment]
Thanks for these replies!

@papandreou The HTML output of Gatsby that I'm running subfont on is the content that is initially visible on the page, without the content that is injected later by React (e.g. the site menu, which is injected via JS after clicking on a hamburger icon). In my case, the menu content is where the issues with missing font characters occurred.  

@Munter The original font was actually loaded before I interacted with the page (the link tag was visible at the end of the body); however, the missing font characters still occurred in the menu when I opened it. So, the issue doesn't seem to be how quickly the original font is loaded (since it is loaded before the interaction occurs). The issue appears to be that the original font is not replacing the subfont version, at least for content dynamically injected using React.

I'm still excited about using `subfont` with Gatsby if we can figure this out!

**Comment by Munter** [comment]
> The original font was actually loaded before I interacted with the page (the Google Fonts link tag was visible at the end of the body); however, the missing font characters still occurred in the menu when I opened it. So, the issue doesn't seem to be how quickly the original font is loaded (since it is loaded before the interaction occurs). The issue appears to be that the original font is not replacing the subfont version, at least for content dynamically injected using React.

The CSS being loaded is not enough in itself to load the font. Fonts only get loaded on demand, and demand is detemined by the browser constructing a render tree where a glyph needs to be rendered with a font that presumably contains it. Since the subfont-created font subset only contains the glyphs for your page content, but not the menu, only the subset is loaded upon page rendering. When you open the menu, that is the time where the render tree calls for loading the original fallback font.

Now, I've done what I can to speed up the load of the google original font by already loading the stylesheet for it, and prewarming the connection to the font serving domain, but you'll still see the font loading roundtrip time as a delay in getting the right font when the menu opens.

This is why I think a prefetch for the fallback font might be in order. Prefetch will load things as soon as the browser thinks there is bandwidth to spare, which usually happens before the page is completely finished loading. This might be an opt-in from the command line, depending on if you value speed or less bandwidth use

**Comment by ooloth** [comment]
> When you open the menu, that is the time where the render tree calls for loading the original fallback font.

> Now, I've done what I can to speed up the load of the google original font by already loading the stylesheet for it, and prewarming the connection to the font serving domain, but you'll still see the font loading roundtrip time as a delay in getting the right font when the menu opens.

Just checking - does this mean that the missing characters in the menu are expected to eventually render with the correct font (after the delay in loading the correct font has passed)? In my case, the missing characters in the menu stay missing and the `font-family` declaration shows the `__subset` version of the font no matter how much time passes. 

I tried leaving the menu open for a few minutes, opening-and-closing the menu a few times, but in all cases, the correct font was never used (i.e. the `font-family` declaration was not updated).

If I've misunderstood, please let me know!

**Comment by papandreou** [comment]
@ooloth, are you able to share the (un`subfont`ed) project?

**Comment by ooloth** [comment]
Thanks again for the replies!

The project is a private repo (it's for a client), but I can create a quick new Gatsby site and recreate the situation. What would be the best way to share the new example project with you? Would a link to its public repo work?

In the meantime, an observation: the modified CSS font-family declarations on my site place the `subfont` version ahead of the original font:

```css
.news-cycle {
    font-family: 'News Cycle__subset', News Cycle,Arial Narrow,sans-serif;
}

.lora {
    font-family: Lora__subset, Lora,Georgia,serif;
}
```

Could this be why the subfont version continues to be used even after the original font has loaded?

**Comment by ooloth** [comment]
Hmm, I also just noticed that the modified CSS font-family declarations above also remove the quotes from multi-word fonts (e.g. `News Cycle` and `Arial Narrow`) that are in my original CSS. Is this quote-removal something that `subset` is (accidentally) doing?

This doesn't relate to my current issue (it's the Lora font that I'm having subsetting issues with), but might be worth investigating as well.

**Comment by Munter** [comment]
I think the missing quotes around the original font name might be a bug. The specifications call for quotes in families with spaces in their names.

It should still fall through to sans-serif though.

You can't trust your browser dev tools here. It just tells you what family applied to the text node in general. It doesn't go to the glyph level, which is where these things are happening. The reason subfont injects the subsets before the original font is that it takes advantage of the fallback through the defined font order of a glyph isn't contained in a higher priority font. This means the original font should load when the menu opens, because that's the time when new glyphs are loaded.

Could you maybe try to put quotes around the news cycle and Arial narrow font names and see if the behavior them switches to load those fonts on demand?

**Comment by ooloth** [comment]
Ah, thanks for the clarifications about the dev tools font-family information.

In my case, the menu only uses the `Lora` font, so modifications to the `News Cycle` declarations won't have any visible effect. (`News Cycle` is also the primary font on the site, so switching the menu to `News Cycle` wouldn't recreate this subsetting issue since all of the characters needed in the menu are also present in the static HTML.)

For whatever reason, it's the characters missing from the `Lora__subset` font are not finding the `Lora` font after the original font has loaded. A puzzle!

**Comment by papandreou** [comment]
I tried a very simple example, where the fallback strategy seems to be working:

```html
<!DOCTYPE html>
<html>
<head>
  <style>
    @font-face {
      font-family: 'Shrikhand';
      font-style: normal;
      font-weight: 400;
      src: local('Shrikhand Regular'), local('Shrikhand-Regular'), url(a8IbNovtLWfR7T7bMJwrA4KU.ttf) format('truetype');
    }

    p { font-family: Shrikhand; }
  </style>
</head>
<body>
  <p>Hello, <span></span>!</p>
  <script>
    setTimeout(() => document.querySelector('span').textContent = 'world', 1000);
  </script>
</body>
</html>
```

The UX is the same before and after applying subfont: `world` appears after a second, and there's no flash of the wrong characters. Probably thanks to the original font being preloaded.

The behavior is the same without the `setTimeout`, or when there's no text at all in the original document.

@ooloth, seems like we will need your help after all. If you find the time to recreate the problem in a fresh repo, it would be easiest to consume if you create a new git repo with the output of gatsby.

We also promise discretion if you're able to share the other site with us in private.

**Comment by ooloth** [comment]
Okay! After some experimenting with a fresh repo, I've managed to clarify that the issue actually relates to the `font-weight` of dynamically-added characters, not to loading the font itself.

The correct font is actually used by all characters in the new content (sorry for my error about that). The reason it looked like the wrong font was being used is that some of the characters in the menu were shown with the wrong font-weight (which I mistook as the wrong font). 

Here is what appears to be happening:
1. If all font-weights used in the dynamically-added content __are also present__ in the static HTML, all characters in the dynamically-added content render correctly.
2. However, __if a new font-weight appears__ in the dynamically-added content: 
    a. __New characters__ that use the new font-weight will display correctly, but
    b. __REUSED characters__ (i.e. characters also present in the static HTML) that use the new font-weight __retain the font-weight used in the static HTML__ (i.e. they do not apply the new font-weight as they should). 

Here is an example repo demonstrating the issue: [https://gitlab.com/ooloth/subfont-demo.git](url)

So, in my case, the real issue is some menu characters which should have a font-weight of `400` are incorrectly rendered using a font-weight of `700` (the only font-weight used for that font in my static HTML).

Adding a hidden `div` containing my menu text to the static HTML apparently fixed my issue not because it included missing characters, but because it used the missing default font-weight of `400`. 

Hopefully that helps clarify things!

**Comment by papandreou** [comment]
@ooloth,

Thanks a lot for the reduced test case. I can reproduce your findings in both Chrome and Firefox. Will look into it!

**Comment by papandreou** [comment]
I think I understand what's going on. This is a case we didn't consider when we implemented the strategy that prepends the `...__subset` font to every `font-family` property that has the font-family being subset as its first entry.

In this case Google's CSS provides a bunch of different variants of the `Lora` font-family, but subfont determines that only the bold one is being used, so it'll only provide a `@font-face` definition of `Lora__subset` for that one. Then when the dynamic content gets injected, the browser will pick `Lora__subset` because even though there's no exact match for the `font-weight`/`font-style` combination, there is a `Lora__subset` that contains the specific characters that are being asked for -- even though they're bold.

This is a consequence of [the "Matching font styles" algorithm](https://www.w3.org/TR/css-fonts-3/#font-style-matching). TL;DR: Only fall back to the next `font-family` if the highest prioritized one does not offer the specific characters that have to be rendered. (The other steps outline a snapping algorithm for which font-weight/font-style combo to fall back to in the absence of an exact match, and because the bold one being the only available one, it will be picked).

I can fix it by injecting a copy of the `font-weight: 400; font-style: normal` `@font-face` with `Lora` changed to `Lora__subset`:

```diff
 @font-face {
   font-display: swap;
   font-family: Lora__subset;
   font-stretch: normal;
   font-style: normal;
   font-weight: 700;
   src: url(Lora-700-d48ffa8dba.woff2) format('woff2'),
     url(Lora-700-7f54e75290.woff) format('woff');
 }
+
+@font-face {
+  font-family: 'Lora__subset';
+  font-style: normal;
+  font-weight: 400;
+  src: local('Lora Regular'), local('Lora-Regular'),
+    url(https://fonts.gstatic.com/s/lora/v12/0QIvMX1D_JOuMwr7Iw.woff2)
+      format('woff2');
+  unicode-range: U+0000-00FF, U+0131, U+0152-0153, U+02BB-02BC, U+02C6, U+02DA,
+    U+02DC, U+2000-206F, U+2074, U+20AC, U+2122, U+2191, U+2193, U+2212, U+2215,
+    U+FEFF, U+FFFD;
+}
```

@Munter, that's the only workaround I can come up with -- I think we'll have to inject definitions of the `__subset` fonts mirroring the `@font-face` definitions of the original font when we don't create a subset of a specific variant because there doesn't appear to be any content requiring it. What do you think?

**Comment by papandreou** [comment]
> When we implement this we could also log an info event that we injected a declaration we couldn't detect the use of, but did so just to be safe.

Yeah, that's a good idea. It'd be a bunch of extra CSS if you've picked many subsets with Google Web Fonts, for example.

> Would there also be a solution in being more selective about which font family properties we prepend the subset name in? Can we detect or record that this specific instance was used in the static analysis result?

I don't think so. The `font`/`font-family` rewriting does not consider the cascade or DOM at all, and I don't think it could. I think our only course of action is to make sure that the rewriting is safe.

Hmm, maybe we can omit echoing the `@font-face` declarations that refer to completely disjoint sets of characters (as determined via `unicode-range`). If the subset only needs to contain enough characters to render "foobar", there's no need to mirror the latin-ext, vietnamese, cyrillic etc. ranges. For once the "matching font styles" algorithm would work in our favor :)

This ought to work whether or not the browser supports `unicode-range`, even with the dynamic CSS from Google Web Fonts in the picture. The browser will still fall back to the non-subset font as soon as it learns that the required characters aren't in the subset font.

**Comment by papandreou** [comment]
The problem with the dynamically injected content being rendered with the wrong `font-weight` should be fixed in subfont 3.6.0.

It'd still be interesting to explore some of the other ideas discussed here for actually getting dynamically injected content included in the subsets, whether automatic or semi-automatic, so I'll leave the issue open.

**Comment by maeh2k** [comment]
I am really impressed with the subfont tool. I didn’t think it could be that simple to get a state of the art font loading strategy running!

However, I have also run into issues with a small GatsbyJS website using the new gatsby subfont plugin ( https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-subfont ) from @KyleAMathews. Subfont generates some subsets and adds a few __subset classes, but most of the elements on the page do not get __subset classes. 

Inspired by @ooloth, I tried to use a workaround and added a hidden div with the critical letters [a-zA-Z] and I manually added __subset classes to all my font-family rules, hoping that subfont would provide the fonts and font-face declaration while I provide the font stack. That didn’t work – it did not add the subset fonts to the head.

**Comment by papandreou** [comment]
@maeh2k, can you share a build of your site as output by gatsby, including your attempt to add the hidden div?

**Comment by maeh2k** [comment]
> @maeh2k, can you share a build of your site as output by gatsby, including your attempt to add the hidden div?

Sure.

Here is the default output from gatsby using the subfont plugin: https://1drv.ms/u/s!AqYU-jg_WCLaoIpx3LMSqTY86KSpIw – the head of the index page contains the links to the subset fonts, but most elements don’t use "__subset" classes; for example the main headline.

Here is the attempted workaround: https://1drv.ms/u/s!AqYU-jg_WCLaoIpwevoXYlgpvyIRYg
A hidden div contains the basic alphabet and I manually added "__subset" css rules by searching and replacing "font-bold" and "font-medium" with "font-bold__subset, font-bold" and "font-medium__subset, font-medium". But the subset fonts are no longer linked in the head.

**Comment by papandreou** [comment]
@maeh2k, as far as I can tell, one of your scripts rerenders most of the page and inject a new stylesheet which hasn't been patched up to include the `__subset` fonts. That's also what's causing the page to flicker briefly right after it has been loaded.

I wonder if this is something gatsby does out of the box when additional rendering takes place at runtime? That would be bad news for us, as there isn't really a good way for subfont to detect and patch that stylesheet, as it's embedded in JavaScript.

**Comment by KyleAMathews** [comment]
Gatsby's server and initial client render are generally identical unless you check for browser apis in the render function (we recommend componentDidMount)

**Comment by KyleAMathews** [comment]
E.g. we don't see a flicker on gatsbyjs.org

**Comment by maeh2k** [comment]
> @maeh2k, as far as I can tell, one of your scripts rerenders most of the page and inject a new stylesheet which hasn't been patched up to include the `__subset` fonts. That's also what's causing the page to flicker briefly right after it has been loaded.

Thank you for looking into it!
I’m new to React/Gatsby, but I wonder if styled-components might inject the new stylesheet.
I am using styled-components v4 with the createGlobalStyle call, where I declare the fonts.
Otherwise, I’m not doing anything fancy in the project (see https://github.com/maeh2k/rhode-medizin ).

**Comment by papandreou** [comment]
@maeh2k, ah, I think it's very likely to be styled-components injecting those styles, then. I don't think there's much we can do about it in subfont itself -- apart from trying to recognize and patch it up in the JavaScript ast, and I'd rather not add such a specific integration.

Just for the experiment you could try adding the `__subset` family to the `font-family` declarations in your styled component definitions?

For example, instead of:

```css
font-family: font-bold,Arial,sans-serif;
```

you'd specify:

```css
font-family: font-bold__subset,font-bold,Arial,sans-serif;
```

It's not ideal by any means (we want subfont to "just work"), but if it works, it'll at least tell us something :)

**Comment by Munter** [comment]
Since styled components v4 is pretty new, this might actually be a missing update to the Gatsby styled components plugin that could fix it

**Comment by maeh2k** [comment]
You may be right that it’s a bit early for styled-components v4 – it’s still in beta and the gatsby plugin may not be v4-ready.

> Just for the experiment you could try adding the `__subset` family to the `font-family` declarations in your styled component definitions?

Not sure how exactly subfont works, but I get the impression that if I manually add "__subset" everywhere, it no longer considers the font in-use and it won’t create the subset fonts and add them to the head. But the workaround appears to work if I only add "__subset" in the styled-components _createGlobalStyle_ method.

It loads the right font immediately, but I still get the annoying flicker at some point – at least when using FireFox; seems better in Chrome. 

Another thing I don’t quite understand and find annoying: when I use --inline-fonts and font-display _block_ or even _optional_, and then throttle Chrome to Slow 3G, it still shows the fallback font for a while and swaps to the webfont later...
The inline CSS looks fine:
```
@font-face{font-display:block;font-family:font-bold__subset;font-stretch:normal;font-style:normal;font-weight:400;src:format("woff2"),url(data:font/woff;base64,
```

**Comment by papandreou** [comment]
@maeh2k,

> Not sure how exactly subfont works, but I get the impression that if I manually add "__subset" everywhere, it no longer considers the font in-use and it won’t create the subset fonts and add them to the head

Ah, yeah, you're right, it won't work. Presumably adding `__subset` everywhere would also affect the server-rendered stylesheet that's present from the beginning, and in that case would indeed affect the tracing code. Never mind :)

**Comment by maeh2k** [comment]
@papandreou FYI, I just published the site with the manual __subset workaround within _createGlobalStyle_ using --inline-fonts and font-display: block: https://rhode-medizin.netlify.com/

At the first glance, it pretty much seems to work – aside from the fact that it will _not_ block the font-display and show the fallback font when I throttle the network. With throttled network speed, the font loading is not there yet.

From time to time when reloading the page, the page will also crash for some reason.
I’m not sure if I introduced that behavior with my subfont workarounds.

**Comment by maeh2k** [comment]
I couldn’t get the data URI version to work. The code looked correct, but for some reason, the resulting subfont did not work.

So I switched to the _preload_ approach. It works with the styled-components workaround and the hidden text (cf. https://rhode-medizin.netlify.com/ ).
Good enough, I guess, though I would have preferred the data URI.

**Comment by papandreou** [comment]
WIP over at https://github.com/Munter/subfont/pull/49, in case someone is up for testing it :)

**Comment by papandreou** [comment]
Released support for `subfont --dynamic` in 4.0.0 :sunglasses:
